org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServerWrapper.java

```diff
@@ -774,12 +787,87 @@ protected IStatus run(IProgressMonitor monitor) {
+	}
+
+	@Override
+	/**
+	 * Runs a request on the language server
+	 *
+	 * @param <T> LS response type
+	 * @param fn Code block that will be supplied the LS in a state where it is guaranteed to have been initialized
+	 *
+	 * @return Async result
+	 */
+	@NonNull
+	public <T> CompletableFuture<T> execute(@NonNull Function<LanguageServer, ? extends CompletionStage<T>> fn) {
+		// Send the request on the dispatch thread, then additionally make sure the response is delivered
+		// on a thread from the default ForkJoinPool. This makes sure the user can't chain on an arbitrary
+		// long-running block of code that would tie up the server response listener and prevent any more
+		// inbound messages being read
+		return executeImpl(fn).thenApplyAsync(Function.identity());
+	}
+
+	/**
+	 * Runs a request on the language server. Internal hook for the LSPexecutor implementations
+	 *
+	 * @param <T> LS response type
+	 * @param fn LS method to invoke
+	 * @return Async result
+	 */
+	@NonNull
+	 <T> CompletableFuture<T> executeImpl(@NonNull Function<LanguageServer, ? extends CompletionStage<T>> fn) {
+		// Run the supplied function, ensuring that it is enqueued on the dispatch thread associated with the
+		// wrapped language server, and is thus guarannteed to be seen in the correct order with respect
+		// to e.g. previous document changes
+		//
+		// Note this doesn't get the .thenApplyAsync(Function.identity()) chained on additionally, unlike
+		// the public-facing version of this method, because we trust the LSPExecutor implementations to
+		// make sure the server response thread doesn't get blocked by any further work
+		return getInitializedServer().thenComposeAsync(fn, this.dispatcher);
+	}
+
+	/**
+	 * Test whether this server supports the requested <code>ServerCapabilities</code>, and ensure
+	 * that it is connected to the document if so.
+	 *
+	 * NB result is a future on this <emph>wrapper</emph> rather than the wrapped language server directly,
+	 * to support accessing the server on the single-threaded dispatch queue.
+	 * @param document Document to connect
+	 * @param filter Constraint on server capabilities
+	 * @return Async result that guarantees the wrapped server will be active and connected to the document. Wraps
+	 * null if the server does not support the requested capabilities or could not be started.
+	 */
+	@NonNull CompletableFuture<@Nullable LanguageServerWrapper> connectIf(@NonNull IDocument document, @NonNull Predicate<ServerCapabilities> filter) {
+		return getInitializedServer().thenCompose(server -> {
+			if (server != null && (filter == null || filter.test(getServerCapabilities()))) {
+				try {
+					return connect(document);
+				} catch (IOException ex) {
+					LanguageServerPlugin.logError(ex);
+				}
+			}
+			return CompletableFuture.completedFuture(null);
+		}).thenApply(server -> server == null ? null : this);
 	}
 
+	@Override
 	/**
 	 * Warning: this is a long running operation
 	 *
```

Comment:
we should move the comment to the interface, shouldn't we? The same applies other methods in this file with now an @Override tag